dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,246 @@
+            => SetCompletionResponse(activity, completions, null, null, completions => $"[{string.Join(", ", completions)}]");
+
+    /// <summary>
+    /// Set the chat completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<ChatMessageContent> completions, int? promptTokens, int? completionTokens)
+        => SetCompletionResponse(activity, completions, promptTokens, completionTokens, ToOpenAIFormat);
+
+    /// <summary>
+    /// Set the chat completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// Token counts will be set to -1 if not provided.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<ChatMessageContent> completions)
+        => SetCompletionResponse(activity, completions, null, null, ToOpenAIFormat);
+
+    # region Private
+    /// <summary>
+    /// Check if model diagnostics is enabled
+    /// Model diagnostics is enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set to true.
+    /// </summary>
+    private static bool IsModelDiagnosticsEnabled()
+    {
+        return s_enableModelDiagnostics || s_enableSensitiveEvents;
+    }
+
+    private static void AddOptionalTags(Activity? activity, PromptExecutionSettings? executionSettings)
+    {
+        if (activity is null || executionSettings?.ExtensionData is null)
+        {
+            return;
+        }
+
+        void TryAddTag(string key, string tag)
+        {
+            if (executionSettings.ExtensionData.TryGetValue(key, out var value))
+            {
+                activity.AddTag(tag, value);
+            }
+        }
+
+        TryAddTag("max_tokens", ModelDiagnosticsTags.MaxToken);
+        TryAddTag("temperature", ModelDiagnosticsTags.Temperature);
+        TryAddTag("top_p", ModelDiagnosticsTags.TopP);
+    }
+
+    /// <summary>
+    /// Convert chat history to a string aligned with the OpenAI format
+    /// </summary>
+    private static string ToOpenAIFormat(IEnumerable<ChatMessageContent> chatHistory)
+    {
+        return $"[{string.Join(
+            ", \n",
+            chatHistory.Select(m => $"{{\"role:\" {m.Role}, \"content:\" {m.Content}}}"))}]";
+    }
+
+    /// <summary>
+    /// Start a completion activity and return the activity.
+    /// The `formatPrompt` delegate won't be invoked if events are disabled.
+    /// </summary>
+    private static Activity? StartCompletionActivity<T>(
+        Uri? endpoint,
+        string modelName,
+        string modelProvider,
+        T prompt,
+        PromptExecutionSettings? executionSettings,
+        Func<T, string> formatPrompt)
+    {
+        if (!IsModelDiagnosticsEnabled())
```

Comment:
```suggestion
        if (!IsModelDiagnosticsEnabled || !s_activitySource.HasListeners)
```

Reply:
`HasListener()` check has been added to `IsModelDiagnosticsEnabled()`.